components/tidb_query/src/codec/mysql/time/datetime.rs

```diff
@@ -0,0 +1,1344 @@
+    }
+}
+
+#[derive(PartialEq, Clone, Copy, Debug)]
+pub enum TimeType {
+    Date,
+    DateTime,
+    TimeStamp,
+}
+
+impl TryFrom<FieldTypeTp> for TimeType {
+    type Error = codec::Error;
+    fn try_from(time_type: FieldTypeTp) -> Result<TimeType> {
+        Ok(match time_type {
+            FieldTypeTp::Date => TimeType::Date,
+            FieldTypeTp::DateTime => TimeType::DateTime,
+            FieldTypeTp::Timestamp => TimeType::TimeStamp,
+            _ => return Err(box_err!("Time does not support field type {}", time_type)),
+        })
+    }
+}
+
+// The common set of methods for `date/time`
+impl Time {
+    /// Returns the hour number from 0 to 23.
+    #[inline]
+    pub fn hour(self) -> u32 {
+        self.get_hour()
+    }
+
+    /// Returns the minute number from 0 to 59.
+    #[inline]
+    pub fn minute(self) -> u32 {
+        self.get_minute()
+    }
+
+    /// Returns the second number from 0 to 59.
+    #[inline]
+    pub fn second(self) -> u32 {
+        self.get_second()
+    }
+
+    /// Returns the number of microseconds since the whole second.
+    pub fn micro(self) -> u32 {
+        self.get_micro()
+    }
+
+    /// Returns the year number
+    pub fn year(self) -> u32 {
+        self.get_year()
+    }
+
+    /// Returns the month number
+    pub fn month(self) -> u32 {
+        self.get_month()
+    }
+
+    /// Returns the day number
+    pub fn day(self) -> u32 {
+        self.get_day()
+    }
+}
+
+mod parser {
+    use super::*;
+    fn bytes_to_u32(input: &[u8]) -> Result<u32> {
+        let mut sum = 0u32;
+        for digit in input {
+            assert(digit.is_ascii_digit())?;
+            sum = sum
+                .checked_mul(10)
+                .and_then(|t| t.checked_add(u32::from(digit - b'0')))
+                .ok_or_else(Error::truncated)?;
+        }
+        Ok(sum)
+    }
+
+    fn digit1(input: &[u8]) -> Result<(&[u8], &[u8])> {
+        let end = input
+            .iter()
+            .position(|&c| !c.is_ascii_digit())
+            .unwrap_or_else(|| input.len());
+        assert(end != 0)?;
+        Ok((&input[end..], &input[..end]))
+    }
+
+    fn space1(input: &[u8]) -> Result<&[u8]> {
+        let end = input
+            .iter()
+            .position(|&c| !c.is_ascii_whitespace())
+            .unwrap_or_else(|| input.len());
+
+        assert(end < input.len())?;
+        Ok(&input[end..])
+    }
+
+    /// We assume that the `input` is trimed and is not empty.
+    fn split_components(input: &str) -> Result<Vec<&[u8]>> {
+        let mut buffer = input.as_bytes();
+
+        assert!(
+            !buffer.is_empty()
+                && !buffer.first().unwrap().is_ascii_whitespace()
+                && !buffer.last().unwrap().is_ascii_whitespace()
+        );
+
+        let mut components = vec![];
+
+        while !buffer.is_empty() {
+            let (mut rest, digits): (&[u8], &[u8]) = digit1(buffer)?;
+
+            components.push(digits);
```

Comment:
It is better to init `components` with `Vec::with_capacity`

Reply:
Thie function will OOM if the string has too many punctuation, because the logic don't limit the num of component part. 
```rust
                // If a punctuation is acquired, move forward the pointer.
                else if rest[0].is_ascii_punctuation() {
                    rest = &rest[1..];
```

Reply:
The logic in the cycle don't limit the num of component parts.